### ISSUE TITLE
GNS3 controller should be a separate entity

### ISSUE THREAD

**Issue by dmfigol** [GNS3/gns3-gui#2541]
Related partially to #2261 and #2256 
I think I brought this up already in some email thread, but the discussion didn't get far.

I believe the current setup is very confusing, especially on the GUI side:
There is a concept of main server which can be local and remote.
There are GNS3 VMs that can be virtual box/vmware or remote.

Now some features will work on one combination and will fail on another. For example, setting running the project in the background will not work if you use local main server and then close the GUI. There are also issues when you try exporting the project from one place to another.
Every so often people come for help to me, and these different options are confusing to people, which results in frustration.

While I understand the reason we have "main server" and remote server or VM, I think it would be better to have a clear separation of entities, which are:

GNS3 controller is responsible for all GNS3 project management side: it knows about the projects, can start and stop them, change settings. It also knows about every compute node and probably it also should be able to synchronize the images available on different compute nodes (on-demand or automatically) to avoid problems.
Controller itself should be something lightweight and standalone (does not require the server like it is now). I would love to use my macbook as a controller without GNS3 server running and my compute would be somewhere remote.
I'd argue that controller should be even separated from GUI and could be run as a daemon.

GNS3 compute node or nodes - can be whatever you want, local, remote or gns3-vm (which is technically a remote server where the hypervisor start/stop is controlled by gui)

While scalability should be supported, the majority of users still use a single compute, hence the user experience should be more focused on them.

With that said, there should be also clear separation of Controller and Compute in GUI.

**Comment by ehlers** [comment]
I like to disagree

> I believe the current setup is very confusing, especially on the GUI side:
There is a concept of main server which can be local and remote.
There are GNS3 VMs that can be virtual box/vmware or remote.

Sure, the documentation could be better and the GUI should be easier to see the concept. But that won't change, if the controller is separate or not.

The controller and the main server are the same, they manage the projects. In contrast the compute server (no matter if GNS3VM, local or remote server) just run the VMs, they have no knowledge of the interconnections between the VMs. Naming both entities "server" may be part of the confusion.

The controller and server functionality is in the same binary (gns3server), because they have a lot of common code and they communicate quite a lot with each other.

> Now some features will work on one combination and will fail on another. For example, setting running the project in the background will not work if you use local main server and then close the GUI. There are also issues when you try exporting the project from one place to another.
Every so often people come for help to me, and these different options are confusing to people, which results in frustration.

I agree, a local controller and a remote one should behave the same. I know, there are currently some differences, especially in the project import/export. For me these are bugs/limitations that should be fixed in future releases. And again, separating the controller from the compute won't magically resolve these issues. The amount of software changes needed is the same for both scenarios.

The local controller/server is normally started and stopped by the GUI. But if you start the local GNS3 server before the GUI, it won't be killed by the GUI at it's exit, it will continue to run. Then your background project should keep on running.

> Controller itself should be something lightweight and standalone (does not require the server like it is now). I would love to use my macbook as a controller without GNS3 server running and my compute would be somewhere remote.

A pure controller wouldn't need much less resources as the combined server and controller. It needs python, some external python modules and the python code. A standalone controller would save just some kilobytes of python code. But as you normally need both entities on a local system, you have two processes with these resource requirements instead of one.

The confusion may come from the GNS3 packaging. The GUI and the server are packaged together, and a lot of additional software is added (dynamips, ubridge, QEMU, VPCS). But strictly speaking, all that is optional. If you just want the controller functionally, you need only the gns3server program.

> I'd argue that controller should be even separated from GUI and could be run as a daemon.

The controller is already a daemon.

> While scalability should be supported, the majority of users still use a single compute, hence the user experience should be more focused on them.

For me, that's another reason against a separation. Then you need two local processes (plus the GUI) instead of one. I don't see, how that makes thinks easier.

To summarize: I agree, that the documentation (which won't be read anyway) and the GUI should be improved. And the whole import/export stuff should be the same on local and remote controller. But other than that, I'm fine with the current state.

**Comment by grossmj** [comment]
Sorry for the late reply, I was travelling. I agree with @ehlers, the only thing we should change here is the way things configured/represented in the GUI and I also agree that a local controller and a remote one should behave the same. However, separating the controller and the compute server code would only make things more complicated and confusing.

**Comment by dmfigol** [comment]
It makes sense, I also tend to agree with @ehlers after his comment.
Can this issue be used to track associated changes that you agreed?
